src/python-frontend/python_converter.cpp

```diff
@@ -346,18 +346,91 @@ locationt
 python_converter::get_location_from_decl(const nlohmann::json &ast_node)
 {
   locationt location;
-  location.set_line(ast_node["lineno"].get<int>());
-  location.set_column(ast_node["col_offset"].get<int>());
+  if (ast_node.contains("lineno"))
+    location.set_line(ast_node["lineno"].get<int>());
+
+  if (ast_node.contains("col_offset"))
+    location.set_column(ast_node["col_offset"].get<int>());
+
   location.set_file(python_filename.c_str());
   location.set_function(current_func_name);
   return location;
 }
 
+symbolt *python_converter::find_function_in_base_classes(
+  const std::string &class_name,
+  std::string symbol_id,
+  std::string method_name,
+  bool is_ctor) const
+{
+  symbolt *func = nullptr;
+
+  // Find class node in the AST
+  auto class_node = json_utils::find_class(ast_json["body"], class_name);
+
+  if (class_node != nlohmann::json())
+  {
+    std::string current_class = class_name;
+    std::string current_func_name = (is_ctor) ? class_name : method_name;
+    // Search for method in all bases classes
+    for (const auto &base_class_node : class_node["bases"])
+    {
+      const std::string &base_class = base_class_node["id"].get<std::string>();
+      if (is_ctor)
+        method_name = base_class;
+
+      std::size_t pos = symbol_id.rfind("@C@" + current_class);
+
+      symbol_id.replace(
+        pos,
+        std::string("@C@" + current_class + "@F@" + current_func_name).length(),
+        std::string("@C@" + base_class + "@F@" + method_name));
```

Comment:
Here, the value given to parameter `symbol_id` is modified in each iteration of the loop. Likely, this should operate on a copy of that value.

Reply:
>I think this will only work for the first iteration.

I guess I got your point now. _method_name_ doesn't need to be updated.
We are searching for the same method across different classes, so only the classname needs to vary.

For example if _main.py_ contains: 
```
class A:
  def foo()

class B(A):
  pass
  
obj = B()
obj.foo()
```
 
After parsing A and B we will have the following symbol for the method foo in the context:  py:main.py@C@A@foo

Given that the obj is of type B, we first look for "py:main.py@C@B@foo" (which doesn't exist), then look for "py:main.py@C@A@foo".

Reply:
That wasn't my point but seems to be valid, too. :)

Consider a class `Z` (one character) and one `XY` (2 characters) to be iterated over in the loop, in this order, and let the current class name be `ABC` (3 characters). What will happen is `sym_id = "py:main.py@C@ABC@F@foo"` is getting replaced by `sym_id = "py:main.py@C@Z@F@foo"`. In the second iteration, the substring `@C@ABC` will not be found anymore in `sym_id`, because after the first iteration the current_class `ABC` has already been replaced with `Z`.

That's why I thought that this should work on a new copy of the original string in each iteration.

Reply:
> In the second iteration, the substring @C@ABC will not be found anymore in sym_id, because after the first iteration the current_class ABC has already been replaced with Z.

current_class is updated at the end of each iteration. So, in the second call to sym_id.rfind("@C@" + current_class);, current_class would hold "Z", and the substring would be found.